materialize-bigquery/binding.go

```diff
@@ -1,225 +1,307 @@
-	// create a file and then directly reference it as a table in a transaction.
-	b.load.extDataConfig = &bigquery.ExternalDataConfig{
-		SourceFormat: bigquery.JSON,
-		Schema:       make([]*bigquery.FieldSchema, 0),
+	if b.Table.RootDocument != nil {
+		document[b.Table.RootDocument.Name()] = string(flowDocument)
 	}
 
-	var pkJoins []string // How to join the main table to the external table based on keys.
-
-	// We use the sqlDriver.NullableTypeMapping to handle null fields for defining schemas
-	// but when using bigquery it does not allow the NOT NULL suffix for external table
-	// columns. We need to get a reference to the base type mapper to leverage it's types.
-	baseTypeMapper := generator.TypeMappings.(sqlDriver.NullableTypeMapping).Inner
+	return document
+}
 
-	// Build the query for loading values from bigquery by joining a google cloud storage file
-	// based table of primary keys against the database.
-	for _, key := range spec.FieldSelection.Keys {
-		var col = tableDef.GetColumn(key)
+// SetWriter will replace whatever writer was attached to this binding
+// If an existing writer is already attached, that writer gets destroyed
+// by calling DestroyWriter() to make sure no data persists on Cloud Storage
+func (b *Binding) SetWriter(ctx context.Context, w *Writer) error {
+	var err error
 
-		colType, err := baseTypeMapper.GetColumnType(col)
-		if err != nil {
-			return nil, err
+	if b.Writer != nil {
+		if err = b.DestroyWriter(ctx); err != nil {
+			return err
 		}
+	}
 
-		b.load.extDataConfig.Schema = append(b.load.extDataConfig.Schema, &bigquery.FieldSchema{
-			Name:     identifierSanitizer(col.Name), // Sanitized name without quoting required
-			Repeated: false,
-			Required: col.NotNull,
-			Type:     bigquery.FieldType(colType.SQLType),
-		})
+	b.Writer = w
 
-		pkJoins = append(pkJoins, fmt.Sprintf("l.%s = r.%s", col.Identifier, col.Identifier))
-	}
+	return nil
+}
 
-	b.load.tempTableName = fmt.Sprintf("%s_load_%d", tempTableNamePrefix, bindingPos)
-
-	if spec.DeltaUpdates {
-		// We should never issue a load query in delta updates mode, so just in case
-		// that happens we'll hopefully produce a reasonable error message with this.
-		b.load.sql = `ASSERT false AS 'Load queries should never be executed in Delta Updates mode.'`
-	} else {
-		// SELECT documents joined with the external table of keys to load
-		b.load.sql = fmt.Sprintf(`
-		SELECT %d, l.%s
-			FROM %s AS l
-			JOIN %s AS r
-			ON %s
-		`,
-			bindingPos,
-			tableDef.GetColumn(spec.FieldSelection.Document).Identifier,
-			tableDef.Identifier,
-			b.load.tempTableName,
-			strings.Join(pkJoins, " AND "),
-		)
+// DestroyWriter clears the writer and delete the underlying
+// asset in Cloud Storage.
+// When this method returns, no writer is associated to this binding
+// and it's required to assign a new one using the SetWriter() method.
+func (b *Binding) DestroyWriter(ctx context.Context) error {
+	defer func() {
+		b.Writer = nil
+	}()
```

Comment:
If we fail to destroy the writer, is it okay to still set it to null? That means we can't retry destroying it I think